src/CommandRunner.cpp

```diff
@@ -196,124 +202,85 @@ class Timer {
-            << ".";
-      }
-      return true;
-    }
-    return false;
-  };
-
-  milliseconds tickPeriod(100);
-  Timer t1(tickPeriod, milliseconds(timeoutInSeconds * 1000));
-  t1.run(waitProcess);
-
-  if (t1.hasTimedOut() && !forceKillRequired) {
-    TASK_LOG(WARNING, loggingMetadata)
-        << "External command took too long to exit. "
-        << "Sending SIGTERM...";
-    if (kill(pid, SIGTERM) == -1) {
-      TASK_LOG(ERROR, loggingMetadata) << "Failed to send SIGTERM: "
-                                       << strerror(errno);
-      forceKillRequired = true;
-    } else {
-      Timer t(tickPeriod, milliseconds(1000));
-      t.run(waitProcess);
-      if (t.hasTimedOut()) {
-        forceKillRequired = true;
-      }
-    }
-  }
-
-  if (forceKillRequired) {
-    TASK_LOG(WARNING, loggingMetadata)
-        << "External command is still running. Sending SIGKILL...";
-    if (kill(pid, SIGKILL) == -1) {
-      TASK_LOG(ERROR, loggingMetadata) << "Failed to kill the command: "
-                                       << strerror(errno);
-      return Error("Command \"" + command +
-                   "\" took too long to execute and SIGKILL failed.");
-    } else {
-      return Error("Command \"" + command + "\" took too long to execute.");
-    }
-  }
-
-  if (exitCode > 0) {
-    return Error("Command \"" + command + "\" exited with return code " +
-                 std::to_string(exitCode) + ".");
-  } else if (signalCode > 0) {
-    return Error("Command \"" + command + "\" exited via signal " +
-                 std::to_string(signalCode) + ".");
+Future<Try<bool>> runCommandWithTimeout(
+    const std::string& theCommand, const std::vector<std::string>& args,
+    unsigned long timeoutInSeconds, const logging::Metadata& loggingMetadata) {
+  // FIXME: remove this
+  vector<string> full_args;
+  full_args.push_back(theCommand);
+  full_args.push_back(args[0]);
+  full_args.push_back(args[1]);
+  full_args.push_back(args[2]);
+  Try<Subprocess> command =
+      subprocess(theCommand, full_args, Subprocess::PATH(args[0]),
+                 Subprocess::PATH(args[1]), Subprocess::PATH(args[2]));
+
+  if (command.isError()) {
+    TASK_LOG(ERROR, loggingMetadata) << "Error launching external command \""
+                                     << theCommand << "\": " << command.error();
+    return Error("launch error");
   }
-  return Nothing();
+  Subprocess process = command.get();
+  return process.status()
+      .then([=](Option<int> status) -> Future<Try<bool>> {
+        if (status.isNone()) {
+          TASK_LOG(ERROR, loggingMetadata)
+              << "Error getting status for external command \"" << theCommand
+              << "\"";
+          return Error("error getting command status");
+        } else if (status.get() != 0) {
+          if (WIFSIGNALED(status.get()) && WTERMSIG(status.get()) != 0) {
+            int signalCode = WTERMSIG(status.get());
+            TASK_LOG(ERROR, loggingMetadata)
```

Comment:
It feels like this whole stuff could have been migrated to dedicated methods so you re-use previous code

Reply:
I'm not confident about that code, it feels like you might break the implementation